### ISSUE TITLE
[Bug] Paste sometimes misses characters on STM32

### ISSUE THREAD

**Issue by cosmikwolf** [marcinbor85/microshell#7]
I have non blocking reads and writes set up over UART using HAL_UART_Transmit_IT and HAL_UART_Receive_IT, and when I paste into the terminal, it sometimes misses characters. 

I have adjusted buffer sizes, and various UART settings to no avail. 

Would be happy to gather some data to help troubleshoot this issue, but I am not entirely sure where to start.

**Comment by cosmikwolf** [comment]
I have also attempted to adjust my terminal settings to add up to a 50ms delay between characters, and it helps, but it will still miss characters maybe 1 / 10 times

**Comment by cosmikwolf** [comment]
awesome library btw :D  I like it alot! I will be trying to implement a command history, so I will do a pull request if I can get it to work!

**Comment by cosmikwolf** [comment]
Ok, I found the cause was that there was something that was blocking in the main loop.  
This may indicate that the ush_service() would run best in an interrupt context on a single threaded MCU

**Comment by marcinbor85** [comment]
Yes, library is not thread-safe, so be careful in multi-thread environment.
Regarding the missing characters, please make sure that in the read/write interfaces You will return 0 in case of not successful write (e.g. buffer overflow) and 1 in case of success (see examples). If You return 0 library will take to write/read attempt on the next service loop.
